src/Expander/TC.hs

```diff
@@ -36,36 +36,34 @@ setTVLevel ptr l = do
   _ <- derefType ptr -- fail if not present
   modifyState $ over (expanderTypeStore . at ptr) $ fmap (set varLevel l)
 
+-- expand outermost constructor to a non-variable, if possible
 normType :: Ty -> Expand Ty
-normType t@(unTy -> TMetaVar ptr) = do
-  tv <- derefType ptr
-  case view varKind tv of
-    Link found -> do
-      t' <- normType (Ty found)
-      setTVKind ptr (Link (unTy t'))
-      return t'
-    _ -> return t
+normType t@(unTy -> TyF (TMetaVar ptr) args) =
+  if null args
+  then do
+    tv <- derefType ptr
+    case view varKind tv of
+      Link found -> do
+        t' <- normType (Ty found)
+        setTVKind ptr (Link (unTy t'))
+        return t'
+      _ -> return t
+  else throwError $ InternalError "type variable cannot have parameters (yet)"
 normType t = return t
 
 normAll :: Ty -> Expand Ty
-normAll t =
-  normType t >>= fmap Ty .
-  \case
-    (Ty (TFun a b)) -> TFun <$> normType a <*> normType b
-    (Ty (TMacro a)) -> TMacro <$> normType a
-    (Ty (TDatatype dt tArgs)) ->
-      TDatatype dt <$> traverse normType tArgs
```

Comment:
This time the omission of `TString` was intentional, as it has no recursive positions on which to call `normType`. The new code handles such argument-less type constructors just fine, and makes it impossible to forget to add new type constructors which do have recursive positions.

Reply:
Great!